날씨 앱 [STEP3-1] bello, roks

KimRoks opened the pull request:
리뷰어: @protocorn93

버드: @JinUng41, @KimRoks

안녕하세요, 콘!!

STEP 3에서는 URLSession을 통해 받아온 데이터를

컬렉션 뷰를 활용하여 화면에 업데이트해주는 작업을 진행하였습니다!

기능적으로는 요구사항을 만족하기는 하나

보완 사항이 더 있을 것이라 판단하여 현재 구현한 부분까지에 대해서 [STEP3-1]로 PR 작성하였습니다.

읽어보시고 코멘트 남겨주시면 감사하겠습니다!!

이번 STEP도 잘 부탁드립니다~
### 요구사항

- [x]  컬렉션 뷰 헤더 뷰의 활용
- [x]  컬렉션 뷰 커스텀 셀의 구현
- [x]  Refresh Control의 활용
- [x]  DateFormatter 등을 활용한 날짜 표현

## 주요 구현 내용

### 디렉터리 구조

```
.
├── Controllers
│   └── ViewController.swift
├── Extensions
│   ├── Bundle.swift
│   ├── DateFormatter.swift
│   └── UIComponents.swift
├── Info.plist
├── Models
│   ├── APIRequest
│   │   ├── Implementations
│   │   │   ├── IconRequest.swift
│   │   │   └── WeatherRequest.swift
│   │   └── Protocols
│   │       └── APIRequest.swift
│   ├── Location
│   │   ├── LocationData.swift
│   │   └── LocationManager.swift
│   ├── Network
│   │   ├── NetworkError.swift
│   │   └── NetworkManager.swift
│   └── Weather
│       ├── Current.swift
│       ├── Forecast.swift
│       └── WeatherType.swift
├── Protocols
│   ├── AlertDisplayable.swift
│   ├── Reusable.swift
│   ├── URLSessionDataTaskProtocol.swift
│   └── URLSessionProtocol.swift
├── Resources
│   ├── AppDelegate.swift
│   ├── Assets.xcassets
│   │   ├── AccentColor.colorset
│   │   │   └── Contents.json
│   │   ├── AppIcon.appiconset
│   │   │   └── Contents.json
│   │   ├── Contents.json
│   │   ├── backgroundImage.imageset
│   │   │   ├── Contents.json
│   │   │   └── diego-ph-5LOhydOtTKU-unsplash.jpg
│   │   └── wallpaper.imageset
│   │       ├── Contents.json
│   │       └── wallpaper.jpg
│   └── SceneDelegate.swift
└── Views
    ├── Base.lproj
    │   └── LaunchScreen.storyboard
    ├── ForecastCell.swift
    ├── WeatherHeaderView.swift
    └── WeatherView.swift
```
### 동작 화면
![동작화면](https://github.com/tasty-code/ios-weather-forecast/assets/113083860/d3a88d25-697c-4efb-a3be-e209913f4368)
## 구현 및 질문

### NetworkManager의 `fetchData()`

기존 코드에서는 Decodable을 따르는 제네릭 타입을 사용하여 `fetchData()`를 구현하였습니다.
Current와 Forecast 타입으로 디코딩 과정이 필요하기 때문이었습니다.

하지만 이미지를 받는 과정에서는 Data 타입을 받아서 UIImage를 생성해야 하기 때문에 fetchData()에서 디코딩 하는 과정은 외부에서 하게 하였습니다.

```swift
// fetchData 함수를 다음과 같이 선언
func fetchData(for request: APIRequest?,
                                 completion: @escaping (Result<Data, NetworkError>)-> Void) {
// 메서드를 호출하는 데에서 디코딩을 거침. (ViewController)
networkManager.fetchData(for: weatherRequest) { [weak self] result in
    switch result {
    case .success(let rawData):
        let currentWeather = try? JSONDecoder().decode(Current.self, from: rawData)
        DispatchQueue.main.async {
            self?.updateHeaderView(with: locationData, currentWeather)
        }
    case .failure(let networkError):
        self?.displayAlert(title: String(describing: networkError))
    }
}
```

### 이미지를 받아오기 위한 과정

`fetchData()`의 구현을 바꾸었기 때문에 이미지를 받는 작업도 해당 메서드를 통해 진행할 수 있었습니다.

```swift
// UICollectionViewDataSource의 cellForItemAt에서..
networkManager.fetchData(for: IconRequest(iconID)) { [weak self] result in
    switch result {
    case .success(let rawData):
        let icon = UIImage(data: rawData)
        DispatchQueue.main.async {
            cell.configureUI(with: listData, icon: icon)
        }
    case .failure(let networkError):
        self?.displayAlert(title: String(describing: networkError))
    }
}
```

### 비대함의 기준

MVC아키텍처의 대표적인 단점이 ViewContoller의 필연적인 비대화라고 알고 있습니다
하지만 이에 대해 명확한 기준이 아직 잡히지 못한 것 같습니다.

예시로 StoryBoard를 사용하여 앱을 만들때 IBOutlet, IBAction등의 코드가 들어가면 비대해지는가?

비즈니스 로직이 얼마나 추가되어야 비대해졌다고 하는것일까?

Controller의 일이 얼마나 어떤 단위로 분산되어야 할까? 등 의 의문이 듭니다.

아주 명확한 정답을 내리기엔 어려운 의문점이겠지만 콘의 의견이 듣고 싶습니다.

### ViewController의 비대해짐과 MVVM 아키텍처의 도입?

구현한 내용을 바탕으로 WeatherViewController가 하는 일이 너무 많지 않나 라는 생각이 들게 되었습니다.

UICollectionViewDataSource, LocationManagerDelegate 이외에 
주입 받은 NetworkManager에게 날씨를 받기 위한 fetchData()를 1차적으로 수행하고
받아온 데이터에서 icon을 다시 받기 위한 2차 네트워크 작업까지 수행하고 있습니다.

따라서 UICollectionViewDataSource나 기타 필수 로직 이외에 ViewModel을 구현하여 로직을 분리하는 것에 대한 Corn의 생각을 듣고 싶습니다.

```mermaid
classDiagram

class ViewController
class WeatherViewModel
class LocationManager
class NetworkManager

LocationManager --> NetworkManager: 위치 업데이트 후 네트워크 실행
NetworkManager --> WeatherViewModel: 네트워크 과정을 거쳐 나온 데이터를 전달? 저장?
WeatherViewModel --> ViewController: 뷰에 업데이트하라고 요청?
```

따라서 만일 코드를 대략적으로 작성해본다면, 다음과 같을 수 있다고 생각이 들었습니다.

```swift
class WeatherViewModel {
    private var currentWeather: Current? {
        didSet {
            // 데이터가 변경되었으니 뷰컨에 업데이트하도록 요청
            updateView()
        }
    }
    private var forecastWeather: Forecast? {
        didSet {
            // 데이터가 변경되었으니 뷰컨에 업데이트하도록 요청
            updateView()
        }
    }
    
    private var networkManager: NetworkManager
    
    init(networkManager: NetworkManager) {
        self.networkManager = networkManager
    }
    
    func processingNetwork() {
        // networkManager에게 일을 시켜 네트워크 작업 후, 데이터를 각각 프로퍼티에 저장
        // didSet이 호출되어 데이터가 변경됨에 따라 updateView()가 호출될 것
    }
    
    private func updateView() {
        // 뷰컨에게 업데이트를 요청하는 로직을 구현
    }
}
```

### 커스텀 셀과 헤더 뷰에 대한 `reuseIdentifier`

이번 스텝에서는 커스텀 셀과 헤더 뷰를 코드로 구현하고 컬렉션 뷰에 `register`하는 과정이 필요하였습니다.

휴먼 에러를 방지하기 위해서 Reusable이라는 프로토콜을 구현하여 타입 속성으로 `reuseIdentifier`를 구현하고 채택하여 사용할 수 있도록 하였습니다.

이와 관련하여 현업에서는 어떻게 사용하고 있는지 궁금증이 생겨 질문드리고 싶습니다.

```swift
protocol Reusable {
    static var reuseIdentifier: String { get }
}

extension Reusable {
    static var reuseIdentifier: String {
        return String(describing: self)
    }
}

// ForecastCell.reuseIdentifier
```

### 앱 자체의 기본 시간을 변경하는 방법

현재 API 호출 시 UTC 기준으로 데이터를 가져오고 있습니다.

때문에, 아래 이미지처럼 한국 시간 16시에 데이터를 호출했지만 이미 지난 시간의 날씨 데이터 또한 가져와 화면에 표시하고 있습니다.

이러한 문제를 해결하고 싶어 AppDelegate에서 많은 시도를 해보았지만 해결하지 못했습니다.

앱 실행 시 현재 위치한 TimeZone에 맞춰 자동으로 세팅되게 하려면 어떤 방법을 사용할 수 있을까요?
<img width="327" alt="스크린샷 2023-12-08 오후 4 01 39" src="https://github.com/tasty-code/ios-weather-forecast/assets/113083860/2a7dce4f-a910-4f01-9caa-547b3acee2f2">

JinUng41 committed:
feat: 코드로 CustomView 구현 및 뷰컨에 매핑

- Assets에 배경 이미지 추가
- CustomView를 코드로 구현 (배경 이미지 뷰, 컬렉션 뷰) + 오토레이아웃 설정
- 뷰컨트롤러의 loadView메서드에서 매핑

KimRoks committed:
feat: 컬렌션 뷰 셀 정의

JinUng41 committed:
feat: Reusable 프로토콜 생성 및 사용

- 컬렉션뷰를 위한 `reuseIdentifier`를 사용할 때,
- 프로토콜로 구현하여 휴먼 에러를 막고자 하였음.

KimRoks committed:
feat: HeaderView 정의

JinUng41 committed:
refactor: UILabel & UIStackView 편의 생성자 구현 및 적용

KimRoks committed:
feat: Refresh Control 구현

JinUng41 committed:
refactor: 컬렉션뷰의 레이아웃 방식 변경

- FlowLayout -> CompositionalLayout
- 헤더뷰에 기본 레이블 텍스트로 하이픈 삽입
- 배경사진 변경
- 레이블 글자색 label로 변경

JinUng41 committed:
feat: 데이터를 받아서 컬렉션뷰에 띄우기

- 뷰컨트롤러에서 데이터를 받았을 때, 컬렉션 뷰의 헤더와 셀에 데이터를 전달하고 업데이트하도록 함.
- 컬렉션뷰의 헤더뷰와 셀 레이아웃 수정
- 새로고침 했을 때, 처음부터 다시 정보를 받아올 수 있도록 함.

JinUng41 committed:
feat: 아이콘을 불러오기 위한 IconRequest 선언 및 폴더 구조 변경

JinUng41 committed:
feat: alert를 위한 프로토콜 선언

- UIViewController를 상속하는 타입이 이 프로토콜을 채택할 시, 기본 제공 함수를 사용하도록 하였음.

JinUng41 committed:
design: 오토레이아웃 변경 및 `final`키워드 추가

JinUng41 committed:
feat: 날씨를 받은 후 이미지를 받기 위한 작업 구현

- ViewController에 마크 주석 추가
- AlertDisplayable을 채택하여 오류에 대한 Alert생성
- 날씨와 이미지를 받고 UI 업데이트

KimRoks committed:
refactor: 전체적인 네이밍 수정

KimRoks committed:
refactor: weatherCollectionView의 은닉화

- WeatherViewController에서 weatherCollcetionView를 직접 참조하지 않도록 수정

protocorn93 commented on WeatherForecast/WeatherForecast/Controllers/WeatherViewController.swift:
```diff
@@ -0,0 +1,181 @@
+    
+    // MARK: - Properties
+
+    private var weatherView: WeatherView!
```
강제 언랩핑을 제거할 수 있는 방법으로 고민 부탁드릴게요~

KimRoks replied:
반영 완료하였습니다!

protocorn93 commented on WeatherForecast/WeatherForecast/Controllers/WeatherViewController.swift:
```diff
@@ -0,0 +1,181 @@
+import UIKit
+import CoreLocation
+
+final class WeatherViewController: UIViewController, AlertDisplayable {
```
AlertDisplayable 구현 및 채택하신 것이 인상 깊네요 👍

JinUng41 replied:
네 감사합니다!! 👍👍

protocorn93 commented on WeatherForecast/WeatherForecast/Controllers/WeatherViewController.swift:
```diff
@@ -0,0 +1,181 @@
+    @objc 
+    private func handleRefreshControl() {
+        locationManager.requestLocation()
+        DispatchQueue.main.async { [weak self] in
+            self?.weatherView.endRefreshing()
+        }
```
refreshControl을 별도 main.async로 감싼 이유가 궁금해요.

KimRoks replied:
refresh control이 작동할때도 UI 업데이트를 하는 과정이라 main큐에서 진행되어야한다고 생각했습니다
하지만 비동기작업 내에서 진행되는 작업이 아니기떄문에 불필요한 과정이었고 수정 완료하였습니다!

protocorn93 commented on WeatherForecast/WeatherForecast/Controllers/WeatherViewController.swift:
```diff
@@ -0,0 +1,181 @@
+            return
+        }
+        
+        guard let headerView = weatherView.findHeaderView() as? WeatherHeaderView else { return }
```
## Optional 

개인적으론 `findHeaderView`보단 `WeatherHeaderView?` 를 반환하는 `headerView` get-only 프로퍼티로 구현하는 것이 좀 더 자연스러워보이긴하네요.

KimRoks replied:
반영 완료하였습니다!

protocorn93 commented on WeatherForecast/WeatherForecast/Controllers/WeatherViewController.swift:
```diff
@@ -0,0 +1,181 @@
+            switch result {
+            case .success(let rawData):
+                let currentWeather = try? JSONDecoder().decode(Current.self, from: rawData)
+                DispatchQueue.main.async {
+                    self?.updateHeaderView(with: locationData, currentWeather)
+                }
```
NetworkManager 내부가 아니라 매번 사용하는 곳에서 completion 클로저를 DispatchQueue.main.async로 감싸는데 completion 클로저를 NetworkManager내부에서 DispatchQueue.main.async로 감싸서 호출하는 것도 고민해보셨을텐데, 외부에서 감싸는 것으로 결정하신 이유가 궁금해요.

KimRoks replied:
최초에 fetchData()에서 제네릭을 채택하다 바꾼 이유와 같습니다.
이는 image를 가져오는 네트워킹 작업에서 범용성이 떨어졌고 따라서 메서드를 수정하게 되었습니다.
매번 사용처에서 DispatchQueue.main.async으로 감싸서 호출하는 이유 또한 
저희가 작성한 fetchData() 메서드는 UI Update만을 위한 메서드가 아닌 
범용적으로 쓰일 수 있는 네트워킹 코드로 쓰이길 바랐기 때문입니다.

protocorn93 replied:
이유에 공감합니다 👍

protocorn93 commented on WeatherForecast/WeatherForecast/Controllers/WeatherViewController.swift:
```diff
@@ -0,0 +1,181 @@
+        }
+    }
+    
+    func notifyLocationErrorAlert() {
+        displayAlert(title: "위치 정보 오류", message: "사용자의 위치 정보를 가져올 수 없습니다.")
+    }
```
개인적으론 ErrorAlert를 메서드에 명시하기보단 didFailedUpdateLocation(error: Error)와 같이 단순히 에러만 전달해주고 에러 처리 방법은 사용부에서 고민하는 것이 좀 더 좋은 구조로 보여지네요.

KimRoks replied:
추후 상세한 에러처리가 가능하도록 Error타입을 인자로 받는 형식으로 수정하였습니다.

protocorn93 commented on WeatherForecast/WeatherForecast/Protocols/Reusable.swift:
```diff
@@ -0,0 +1,18 @@
+
+import Foundation
+
+protocol Reusable {
+    static var reuseIdentifier: String { get }
+}
+
+extension Reusable {
+    static var reuseIdentifier: String {
+        return String(describing: self)
+    }
+}
```
좋네요! 많은 회사, 많은 팀에서도 이렇게 구핸해놓긴해요!

JinUng41 replied:
네 감사합니다!

protocorn93 commented on WeatherForecast/WeatherForecast/Views/ForecastCell.swift:
```diff
@@ -0,0 +1,71 @@
+
+final class ForecastCell: UICollectionViewCell, Reusable {
+    
+    private lazy var dateLabel: UILabel = {
+        let label = UILabel(textAlignment: .left)
+        addSubview(label)
+        
+        return label
+    }()
+    
+    private lazy var temperatureLabel: UILabel = {
+        let label = UILabel(textAlignment: .right)
+        
+        return label
+    }()
+    
+    private lazy var weatherIconView: UIImageView = {
+        let imageView = UIImageView()
+        imageView.contentMode = .scaleAspectFit
+        
+        return imageView
+    }()
```
이 세 프로퍼티는 lazy var로 선언되지 않아도 무방하지 않나요?

KimRoks replied:
저희가 lazy var 로 통일했던 이유는
addSubViews가 필요한 컴포넌트와 통일성을 위한 코드였습니다.
lazy가 불필요한 부분에 대해서 수정을 진행하였으나
콘께서 의도하시는 내용이 정확히 어떤것인지 궁금합니다
일전에 Codable -> Decodable로 수정한 예시처럼 
코드의 명확한 의미전달을 위해서 일까요?

protocorn93 replied:
네네 저는 단순히 코드 구조의 통일성 보다는 명확한 사용 이유가 있길 바랬어요

protocorn93 commented on WeatherForecast/WeatherForecast/Views/WeatherHeaderView.swift:
```diff
@@ -0,0 +1,98 @@
+
+import UIKit
+
+final class WeatherHeaderView: UICollectionReusableView, Reusable {
```
마찬가지로 불필요한 lazy var 선언은 제거되면 좋겠네요!

KimRoks replied:
수정하였습니다!

protocorn93 commented on WeatherForecast/WeatherForecast/Views/WeatherView.swift:
```diff
@@ -0,0 +1,100 @@
+
+import UIKit
+
+final class WeatherView: UIView {
```
WeatherView를 별도로 선언하여 뷰 구성에 대한 역할을 VC로부터 분리해내신 점 좋네요! 👍

JinUng41 replied:
네 감사합니다! 👍👍

protocorn93 commented on WeatherForecast/WeatherForecast/Views/WeatherView.swift:
```diff
@@ -0,0 +1,100 @@
+        return collectionView
+    }()
+    
+    init(delegate: UICollectionViewDataSource) {
```
`UICollectionViewDataSource`도 `WeatherView`가 채택해도 되지 않을까요? WeatherViewDelegate 같은 친구를 선언해서 필요한 데이터나 인터페이스는 외부에서 적절히 제공 받아도 되구요.

JinUng41 replied:
`WeatherView`가 `UICollectionViewDataSource`를 채택하게 되면, 데이터를 받아 셀에 반영하는 것을 뷰 클래스에서 담당하게 됩니다. 그렇게 되면 뷰의 책임? 역할?에 어긋난다고 생각되는데요. 혹시 자세한 설명을 들을 수 있을까요?

protocorn93 replied:
우선 저는 ViewController/View 둘 모두 UI 즉 뷰로 바라보아요. 

데이터 가공이 아닌 적용이기 때문에 저는 이것은 뷰 계층의 영역이라고 생각해요.

protocorn93 replied:
네트워크 응답의 결과로 받은 값을 UI에 보여주기 위해 가공하는 역할은 비즈니스 로직 객체 혹은 프레젠테이션 로직 객체에 위임할 수 있지만

이를 적용하는 것은 UI가 직접하는 것이 맞다고보여져요.

protocorn93 replied:
아무래도 UICollectionViewDataSource를 활용하다보니 이 부분의 경계가 애매해보이시는 것 같아요. DiffableDataSource를 활용해보다면 계층의 구분을 좀 더 명확하게 가져가보실 수 있을거에요. 이에 대해선 이후에 한번 학습해보시고 적용해보셔도 좋을 것 같아요.

protocorn93 reviewed (changes requested):
두 분 모두 고생하셨습니다. 

## 질문 사항

> 비대함의 기준

저와 제가 속한 팀에서 VC의 비대함은 UI 구성과 별개의 로직이 들어가는 것을 모두 비대함으로 정의하고 있어요. 모든 비즈니스 로직은 VC 외의 별도 객체로 분리하고 처리하고 있어요. 버튼을 눌리면 `someObject.handleButtonClick`과 같이 액션을 전달하고 있구요. 

이런식으로 비즈니스 로직은 모두 외부 객체에 위임하고 있어요. 결과는 주로 closure 주입을 통해 전달하거나 Delegate로 전달하긴해요. 

이것의 결과물로 MVVM과 같은 다양한 아키텍처가 도입될 수 있겠죠.

> 커스텀 셀과 헤더 뷰에 대한 reuseIdentifier

네네! 많은 팀에서 이렇게 구현하고 있어요. 

> 현재 API 호출 시 UTC 기준으로 데이터를 가져오고 있습니다.

서버가 UTC 기준이라면 호출할 때 현재 디바이스의 시간을 UTC 기준으로 변환해서 호출해주는 방법밖에 없을 것 같아요. 응답도 UTC 기준으로 온다면 이를 디바이스 시간을 기준으로 맞춰 포맷팅이 필요하겠죠! Locale 관련해서도 한번 학습해보시면 좋을 것 같아요, 이는 iOS에 국한되는 것이 아니기에 학습해보시면 앞으로도 많은 도움이 되실 수 있어요, 

- https://www.44bits.io/ko/keyword/locale

KimRoks committed:
refactor: 3-1 피드백 반영

- 강제 옵셔널 바인딩 제거
- 불필요한 lazy 키워드 제거
- LocationUpdateDelegate 수정
- Refresh Conrol 작업이 불필요한 Dispatch Queue를 사용하지 않도록 수정

protocorn93 commented on WeatherForecast/WeatherForecast/Controllers/WeatherViewController.swift:
```diff
@@ -10,16 +10,16 @@ import CoreLocation
-
-    private var weatherView: WeatherView!
+    
+    private var weatherView: WeatherView?
```
`weatherView`는 `loadView` 시점에 할당이 되어요. `loadView`는 기본적으로 lazy하게 뷰를 할당하기 때문에 `weatherView`는 `lazy var`를 통해 생성해주어도 되요.

JinUng41 replied:
넵 수정하였습니다!!

JinUng41 committed:
refactor: weatherView를 lazy하게 생성

JinUng41 committed:
refactor: lazy 생성되는 weatherView에 따른 코드 수정

- 불필요한 옵셔널 바인딩 및 체이닝 삭제
